### PR TITLE
Specialized vector rand! for many distributions

### PR DESCRIPTION
**Pull request by quildtide**
The much-larger sequel to #1269 and #1680

All claims tested on julia 1.11.0-rc1, Windows 10, AMD Ryzen 9 5900X, RTX 3080

I was originally planning on making the normal dispatch path faster for all univariate distributions on `rand!`, but it turns out that the current dispatch path is  probably about as fast as we will ever get with a generic rand -> rand! fallback path. I tested a few alternatives and they were all similar or slower.

The root cause of why this dispatch path is slower than specialized vector methods is because Julia's base `rand`/`randn`/`randexp` functions have `rand!` specializations for both Xoshiro and Mersenne that perform much faster than calling scalar `rand` many times.  [Xoshiro actually has a separate SIMD implementation in Random.jl](https://github.com/JuliaLang/julia/blob/master/stdlib/Random/src/XoshiroSimd.jl) [and both Mersenne and Xoshiro branch based off of vector length to choose optimal algorithms](https://github.com/JuliaLang/julia/blob/master/stdlib/Random/src/normal.jl#L250).

Creating specialized `rand!` methods yields some pretty impressive performance gains as a result. I plan on continuing further with this, but this pull request contains:

## Vectorized `rand!` specializations for:

- Exponential (~67% faster)
- ~~Laplace (\~185% faster)~~
- Pareto (~86% faster)
- ~~InverseGaussian (\~102% faster)~~
- LogitNormal (~46% faster)
- LogNormal (~234% faster)
- NormalCanon (~108% faster)
- ~~PGeneralizedGaussian (\~60% faster)~~

Benchmarking was done in a very ad hoc way; I grabbed 3 values from `@time`, excluding any runs that included compilation or GC time. This is a small sample size, but results were pretty consistent and I don't really feel a need to prove significance when they're all consistently >40% faster.

### Improved GPU compatibility

I don't have any non-CUDA GPUs to test with, so testing has been CUDA-only, but these things are probably true for other GPU types too.

CUDA.jl does not like the for-loop in the generic fallback `rand!` dispatch path in Distributions.jl. However, specialized dispatch paths can help alleviate this issue.

These distributions now have functional out-of-the-box CUDA.jl support for `rand!`:
- LogitNormal
- LogNormal
- NormalCanon

These distributions would probably work with the addition of a `randexp` implementation in CUDA.jl:
- Exponential
- Pareto
- ~~Laplace~~

~~`rand!(::PGeneralizedGaussian)` is dependent on `rand(::Gamma)`. CUDA.jl does not like functions that call `rand(::Gamma)`.~~

~~I'm not sure why `rand(::InverseGaussian)` isn't working. I'll have to revisit that later.~~

### Improved Dagger.jl compatibility

The current default fallback `rand!` methods perform so poorly on Dagger.jl that they cannot even really be benchmarked. This is mostly due to issues on the Dagger.jl side that are currently being addressed. I only did a tiny bit of testing with these new methods, but they do perform better than the generic fallbacks.

### Minor adjustments to scalar `rand` methods:

While I was implementing the specialized `rand!` methods, I noticed opportunities to make small changes to some scalar `rand` methods that would, in theory, yield small performance or accuracy benefits. In practice, none of these changes seem to have significant observable changes in speed:
- LogitNormal uses `muladd` now
- Lognormal uses `muladd` now
- Normal uses `muladd` now
- PGeneralizedGaussian calls `rand(rng, Bool)` instead of `rand(rng) < 0.5` now

## Additional bloating for the test suite

One negative drawback to increased specialization of methods is that there's a larger surface for bugs to emerge. This is doubly so when I'm also making small adjustments to the scalar `rand` methods.

I compensated by adding new functionality to `test_distr` and `test_samples` (enabled by keyword) that allows these functions to call the scalar `rand` functions multiple times. I also added `test_distr` entries for all of the distributions that I've added new methods for.

Unfortunately, I'm pretty sure a lot of the testing I added also duplicates work done by existing tests. :shrug:

I didn't remove any existing tests because I was unsure exactly which parts were redundant. I think code coverage probably went up a bit as a side effect.

I also had to make minor tolerance adjustments to 2 tests.

## Future Plans

A lot of distributions are dependent on `rand(::Gamma)`, which is non-trivial to translate to a vectorized form. Accomplishing this would probably enable out-of-the-box GPU support and large performance gains for all dependent distributions.

Most of the distributions addressed in this pull request were picked because I thought they would be trivial or easy to implement, but there's still a ton of others remaining in this category.

### CODE REVIEW

**Comment by quildtide** [comment]
This pull request will probably cause a repeat of #1687 on a larger scale, but I don't think it's really an issue, as this is consistent with Julia behavior. Weird trivia example regarding Random.jl:

`rand(Xoshiro(123), 7) == rand(Xoshiro(123), 7)`

`rand(Xoshiro(123), 7) != rand(Xoshiro(123), 8)[1:7]`

**Comment by quildtide** [comment]
@devmotion Could I get this reviewed if the scope of the pull request is not too wide?

File: src/univariate/continuous/pgeneralizedgaussian.jl [hunk 1]
```diff
@@ -141,9 +141,20 @@ function rand(rng::AbstractRNG, d::PGeneralizedGaussian)
...
     z = d.α * rand(rng, g)^inv_p
-    if rand(rng) < 0.5
+    if rand(rng, Bool)
         return d.μ - z
     else
         return d.μ + z
     end
 end
+
+function rand!(rng::AbstractRNG, d::PGeneralizedGaussian, A::AbstractArray{<:Real})
```
Comment (devmotion) [inline src/univariate/continuous/pgeneralizedgaussian.jl:151]: This function suffers from the same problems as mentioned above, can you remove it from the PR?

File: src/univariate/continuous/pareto.jl [hunk 2]
```diff
@@ -112,6 +112,12 @@ quantile(d::Pareto, p::Real) = cquantile(d, 1 - p)
 
 rand(rng::AbstractRNG, d::Pareto) = d.θ * exp(randexp(rng) / d.α)
 
+rand!(rng::AbstractRNG, d::Pareto, A::AbstractArray{<:Real}) =
+    # A .* 1/d.α is faster than A ./ d.α, but less exact
+    # fast-math style optimization; should be fine for PRNG
+    A .=  d.θ .* exp.(randexp!(rng, A) .* (1/d.α))
+
+
```
Comment (devmotion) [inline src/univariate/continuous/pareto.jl:120]: I think using `randexp!` is an obvious improvement but I'm not sure whether `1/...` and the inconsistency between `rand` and `rand!` is worth it (generally `1/...` can easily lead to undesired promotions but I don't think that's a problem here). I'd use something like

```suggestion
xval(d::Pareto, z::Real) = d.θ * exp(z / d.α)

rand(rng::AbstractRNG, d::Pareto) = xval(d, randexp(rng))
function rand!(rng::AbstractRNG, d::Pareto, A::AbstractArray{<:Real})
    randexp!(rng, A)
    map!(Base.Fix1(xval, d), A, A)
    return A
end

```

File: src/univariate/continuous/normalcanon.jl [hunk 3]
```diff
@@ -89,6 +89,12 @@ invlogccdf(d::NormalCanon, lp::Real) = xval(d, norminvlogccdf(lp))
 
 rand(rng::AbstractRNG, cf::NormalCanon) = cf.μ + randn(rng) / sqrt(cf.λ)
 
+rand!(rng::AbstractRNG, cf::NormalCanon, A::AbstractArray{<:Real}) =
+    # A .* inv(sqrt(cf.λ)) is faster but less accurate than A ./ sqrt(cf.λ)
+    # in floating point. This is a fast-math style optimization.
+    # It shouldn't be a problem for PRNG.
+    A .= muladd.(randn!(rng, A), inv(sqrt(cf.λ)), cf.μ)
```
Comment (devmotion) [inline src/univariate/continuous/normalcanon.jl:96]: I think using `randn!` os an obvious improvement but I'm not sure whether `inv` + `muladd` and the inconsistency between `rand` and `rand!` is worth it (generally `inv` can easily lead to undesired promotions but I don't think that's a problem here). I'd use something like


```suggestion
function rand!(rng::AbstractRNG, cf::NormalCanon, A::AbstractArray{<:Real})
    randn!(rng, A)
    A .= cf.μ .+ A ./ sqrt(cf.λ)
    return A
end
```

File: src/univariate/continuous/exponential.jl [hunk 4]
```diff
@@ -107,6 +107,8 @@ cf(d::Exponential, t::Real) = 1/(1 - t * im * scale(d))
 #### Sampling
 rand(rng::AbstractRNG, d::Exponential) = xval(d, randexp(rng))
 
+rand!(rng::AbstractRNG, d::Exponential, A::AbstractArray{<:Real}) =
+    A .= xval.(d, randexp!(rng, A))
```
[the remark this answers is not in the document]
Comment (quildtide) [reply src/univariate/continuous/exponential.jl:111]: I recall testing something similar before with `map!`, and it was slightly slower and caused allocations, but your version has no allocations and is slightly faster.

File: src/univariate/continuous/inversegaussian.jl [hunk 5]
```diff
@@ -169,8 +169,23 @@ function rand(rng::AbstractRNG, d::InverseGaussian)
...
+function rand!(rng::AbstractRNG, d::InverseGaussian, A::AbstractArray{<:Real})
+    # Based off of the non-vectorized code
+    μ, λ = params(d)
+    Z = V = W = X1 = A # prevents extra heap allocs
+    randn!(rng, A)
+    V .*= Z
+    W .*= μ
+    X1 .= μ .+ μ / (2λ) .* (W .- sqrt.(W .* (4λ .+ W)))
+    A .= ifelse.( # TODO Can we avoid heap-allocing a whole float vec?
+        rand(rng, size(A)...) .>= μ ./ (μ .+ X1),
```
[the remark this answers is not in the document]
Comment (quildtide) [reply src/univariate/continuous/inversegaussian.jl:181]: You know, that explains why this wasn't working on CUDA earlier; hadn't thought of that.

**Comment by quildtide** [comment]
@devmotion

File: test/testutils.jl [hunk 6]
```diff
@@ -28,8 +28,9 @@ end
 # testing the implementation of a discrete univariate distribution
 #
 function test_distr(distr::DiscreteUnivariateDistribution, n::Int;
-                    testquan::Bool=true)
-
+                    testquan::Bool=true, rng::AbstractRNG=MersenneTwister(),
+                    test_scalar_rand::Bool=false)
```
Comment (devmotion) [inline test/testutils.jl:32]: I'd like to keep the changes a bit more minimal - I think we can continue working with the `MersenneTwister` and test the the scalar `rand` method by default?
```suggestion
                    testquan::Bool=true)
```

File: test/testutils.jl [hunk 7]
```diff
@@ -28,8 +28,9 @@ end
...
+    println("    testing $(distr)")
```
Comment (devmotion) [inline test/testutils.jl:33]: I think tests are already quite verbose? If needed, they can also be grouped/wrapped in a `@testset`.
```suggestion
```

File: test/testutils.jl [hunk 8]
```diff
@@ -40,7 +41,12 @@ function test_distr(distr::DiscreteUnivariateDistribution, n::Int;
 
     test_stats(distr, vs)
     test_samples(distr, n)
-    test_samples(distr, n, rng=MersenneTwister())
+    test_samples(distr, n, rng=rng)
+    if test_scalar_rand
+        xs = test_samples(distr, n; call_scalar = true)
+        xs = test_samples(distr, n, rng=rng, call_scalar = true)
+    end
+
```
Comment (devmotion) [inline test/testutils.jl:49]: I think we can just revert to
```suggestion
    test_samples(distr, n, rng=MersenneTwister())
```

File: test/testutils.jl [hunk 9]
```diff
@@ -67,7 +73,9 @@ end
 # testing the implementation of a continuous univariate distribution
 #
 function test_distr(distr::ContinuousUnivariateDistribution, n::Int;
-                    testquan::Bool=true, rng::AbstractRNG=MersenneTwister(123))
+                    testquan::Bool=true, rng::AbstractRNG=MersenneTwister(123),
+                    test_scalar_rand::Bool=false)
```
Comment (devmotion) [inline test/testutils.jl:77]:
```suggestion
                    testquan::Bool=true, rng::AbstractRNG=MersenneTwister(123))
```

File: test/testutils.jl [hunk 10]
```diff
@@ -67,7 +73,9 @@ end
...
+    println("    testing $(distr)")
```
Comment (devmotion) [inline test/testutils.jl:78]:
```suggestion
```

File: test/testutils.jl [hunk 11]
```diff
@@ -82,6 +90,13 @@ function test_distr(distr::ContinuousUnivariateDistribution, n::Int;
     allow_test_stats(distr) && test_stats(distr, xs)
     xs = test_samples(distr, n, rng=rng)
     allow_test_stats(distr) && test_stats(distr, xs)
+    if test_scalar_rand
+        xs = test_samples(distr, n; call_scalar = true)
+        allow_test_stats(distr) && test_stats(distr, xs)
+        xs = test_samples(distr, n, rng=rng, call_scalar = true)
+        allow_test_stats(distr) && test_stats(distr, xs)
+    end
+
```
Comment (devmotion) [inline test/testutils.jl:99]:
```suggestion
```

File: test/testutils.jl [hunk 12]
```diff
@@ -101,8 +116,8 @@ function test_samples(s::Sampleable{Univariate, Discrete},      # the sampleable
                       n::Int;                                   # number of samples to generate
                       q::Float64=1.0e-7,                        # confidence interval, 1 - q as confidence
                       verbose::Bool=false,                      # show intermediate info (for debugging)
-                      rng::Union{AbstractRNG, Missing}=missing) # add an rng?
-
+                      rng::Union{AbstractRNG, Missing}=missing,  # add an rng?
+                      call_scalar::Bool=false)                  # directly scall the scalar rand(d) instead
```
Comment (devmotion) [inline test/testutils.jl:120]:
```suggestion
                      rng::Union{AbstractRNG, Missing}=missing) # add an rng?
```

File: test/testutils.jl [hunk 13]
```diff
@@ -146,14 +161,30 @@ function test_samples(s::Sampleable{Univariate, Discrete},      # the sampleable
...
+        end
         Random.seed!(1234)
-        samples = rand(s, n)
-        Random.seed!(1234)
-        samples2 = rand(s, n)
+        samples2 = if !call_scalar
+            rand(s, n)
+        else
+            map((_) -> rand(s), 1:n)
+        end
```
Comment (devmotion) [inline test/testutils.jl:175]: Can we introduce `samples3` and `samples4` unconditionally, that sample with `rand`? And then check their equality (`samples3 == samples4`) and the properties below as well?

File: test/testutils.jl [hunk 14]
```diff
@@ -146,14 +161,30 @@ function test_samples(s::Sampleable{Univariate, Discrete},      # the sampleable
...
+        samples = if !call_scalar
+            rand(rng, s, n)
+        else
+            map((_) -> rand(rng, s), 1:n)
+        end
+        samples2 = if !call_scalar
+            rand(rng2, s, n)
+        else
+            map((_) -> rand(rng2, s), 1:n)
+        end
```
Comment (devmotion) [inline test/testutils.jl:187]: Same here, test these alternatives unconditionally?

File: test/testutils.jl [hunk 15]
```diff
@@ -180,8 +211,8 @@ function test_samples(s::Sampleable{Univariate, Discrete},      # the sampleable
 end
 
 test_samples(distr::DiscreteUnivariateDistribution, n::Int;
-             q::Float64=1.0e-6, verbose::Bool=false, rng=missing) =
-    test_samples(distr, distr, n; q=q, verbose=verbose, rng=rng)
+             q::Float64=1.0e-6, verbose::Bool=false, rng=missing, kwargs...) =
+    test_samples(distr, distr, n; q=q, verbose=verbose, rng=rng, kwargs...)
```
Comment (devmotion) [inline test/testutils.jl:215]:
```suggestion
             q::Float64=1.0e-6, verbose::Bool=false, rng=missing) =
    test_samples(distr, distr, n; q=q, verbose=verbose, rng=rng)
```

File: test/testutils.jl [hunk 16]
```diff
@@ -191,7 +222,8 @@ function test_samples(s::Sampleable{Univariate, Continuous},    # the sampleable
                       nbins::Int=50,                            # divide the main interval into nbins
                       q::Float64=1.0e-6,                        # confidence interval, 1 - q as confidence
                       verbose::Bool=false,                      # show intermediate info (for debugging)
-                      rng::Union{AbstractRNG, Missing}=missing) # add an rng?
+                      rng::Union{AbstractRNG, Missing}=missing, # add an rng?
+                      call_scalar::Bool=false)                  # directly scall the scalar rand(d) instead
```
Comment (devmotion) [inline test/testutils.jl:226]:
```suggestion
                      rng::Union{AbstractRNG, Missing}=missing) # add an rng?
```

File: test/testutils.jl [hunk 17]
```diff
@@ -284,8 +332,8 @@ function test_samples(s::Sampleable{Univariate, Continuous},    # the sampleable
     return samples
 end
 
-test_samples(distr::ContinuousUnivariateDistribution, n::Int; nbins::Int=50, q::Float64=1.0e-6, verbose::Bool=false, rng=missing) =
-    test_samples(distr, distr, n; nbins=nbins, q=q, verbose=verbose, rng=rng)
+test_samples(distr::ContinuousUnivariateDistribution, n::Int; nbins::Int=50, q::Float64=1.0e-6, verbose::Bool=false, rng=missing, kwargs...) =
+    test_samples(distr, distr, n; nbins=nbins, q=q, verbose=verbose, rng=rng, kwargs...)
```
Comment (devmotion) [inline test/testutils.jl:336]:
```suggestion
test_samples(distr::ContinuousUnivariateDistribution, n::Int; nbins::Int=50, q::Float64=1.0e-6, verbose::Bool=false, rng=missing) =
    test_samples(distr, distr, n; nbins=nbins, q=q, verbose=verbose, rng=rng)
```

File: test/testutils.jl [hunk 18]
```diff
@@ -246,14 +278,30 @@ function test_samples(s::Sampleable{Univariate, Continuous},    # the sampleable
     # generate samples using RNG passed or default RNG
     # we also check reproducibility
     if rng === missing
+        Random.seed!(1234)    
+        samples = if !call_scalar
+            rand(s, n)
+        else
+            map((_) -> rand(s), 1:n)
+        end
```
Comment (devmotion) [inline test/testutils.jl:286]: Run these additional branches unconditionally?

File: test/univariate/continuous/exponential.jl [hunk 19]
```diff
@@ -8,3 +7,18 @@
...
+test_cgf(Exponential(10  ), (0.08, -1, -100f0, -1e6))
+
+# Sampling Tests
+@testset "Exponential sampling tests" begin
+    for d in [
+        Exponential(1),
+        Exponential(0.91),
+        Exponential(10)
+    ]
+        test_distr(d, 10^6, test_scalar_rand = true)
```
Comment (devmotion) [inline test/univariate/continuous/exponential.jl:22]:
```suggestion
        test_distr(d, 10^6)
```

File: test/univariate/continuous/logitnormal.jl [hunk 20]
```diff
@@ -64,3 +64,12 @@ end
     @test convert(LogitNormal{Float32}, d) === d
     @test typeof(convert(LogitNormal{Float64}, d)) == typeof(LogitNormal(2,1))
 end
+
+@testset "Logitnormal Sampling Tests" begin
+    for d in [
+        LogitNormal(-2, 3),
+        LogitNormal(0, 0.2)
+    ]
+        test_distr(d, 10^6, test_scalar_rand = true)
```
Comment (devmotion) [inline test/univariate/continuous/logitnormal.jl:73]:
```suggestion
        test_distr(d, 10^6)
```

File: test/univariate/continuous/lognormal.jl [hunk 21]
```diff
@@ -314,3 +314,17 @@ end
...
+    for d in [
+            LogNormal()
+            LogNormal(1.0)
+            LogNormal(0.0, 2.0)
+            LogNormal(1.0, 2.0)
+            LogNormal(3.0, 0.5)
+            LogNormal(3.0, 1.0)
+            LogNormal(3.0, 2.0)
+        ]
+        test_distr(d, 10^6, test_scalar_rand = true)
```
Comment (devmotion) [inline test/univariate/continuous/lognormal.jl:328]:
```suggestion
        test_distr(d, 10^6)
```

File: test/univariate/continuous/normalcanon.jl [hunk 22]
```diff
@@ -0,0 +1,10 @@
+# Sampling Tests
+@testset "NormalCanon sampling tests" begin
+    for d in [
+        NormalCanon()
+        NormalCanon(-1.0, 2.5)
+        NormalCanon(2.0, 0.8)
+    ]
+        test_distr(d, 10^6, test_scalar_rand = true)
```
Comment (devmotion) [inline test/univariate/continuous/normalcanon.jl:8]:
```suggestion
        test_distr(d, 10^6)
```

File: test/univariate/continuous/pareto.jl [hunk 23]
```diff
@@ -0,0 +1,10 @@
+@testset "Pareto Sampling Tests" begin
+    for d in [
+            Pareto()
+            Pareto(2.0)
+            Pareto(2.0, 1.5)
+            Pareto(3.0, 2.0)
+        ]
+        test_distr(d, 10^6, test_scalar_rand = true)
```
Comment (devmotion) [inline test/univariate/continuous/pareto.jl:8]:
```suggestion
        test_distr(d, 10^6)
```

**Review by devmotion — COMMENTED** [review]
Can you please add newline characters at the end of the newly added (and modified) pages? Thank you!

**Comment by quildtide** [comment]
@devmotion Requested changes have been made.

**Review by devmotion — APPROVED** [review]
Thank you, @quildtide!

**Pull request merged**